linux/mcompile/keymap/keymap.cpp

```diff
@@ -109,355 +472,330 @@ void Split_US_To_3D_Vector(v_str_3D &all_US,v_str_1D completeList) {
+  for(int i =0; i< (int) All_Vector[1].size();i++) {
 
-  for(int i =1; i< (int) All_Vector[1].size()-1;i++)
-  {
-    // get key name US stored in [0][i][0] and copy to name in other-block[1][i][0]
+    // get key name US stored in [0][i][0] and copy to name in "underlying"-block[1][i][0]
     All_Vector[1][i][0] = All_Vector[0][i][0];
 
-    // write this value to 3D- Vector
-    All_Vector[1][i][0+1] = GetKeyvalsFromKeymap(keymap,stoi(All_Vector[1][i][0]),0);   //shift state: unshifted:0
-    All_Vector[1][i][1+1] = GetKeyvalsFromKeymap(keymap,stoi(All_Vector[1][i][0]),1);   //shift state: shifted:1
-    //printf("Keycodes US->Other:   %d(US): %s %s ---- (other):%s,  %s,  %s  \n",stoi(All_Vector[1][i][0]),All_Vector[0][i][1].c_str(),All_Vector[0][i][2].c_str(),All_Vector[1][i][1].c_str(),All_Vector[1][i][2].c_str(),All_Vector[1][i][3].c_str());  
+    // get Keyvals of this key and copy to unshifted/shifted in "underlying"-block[1][i][1] / block[1][i][2]
+    All_Vector[1][i][0+1] = KMX_get_KeyValUnderlying_From_KeyCodeUnderlying(keymap,All_Vector[0][i][0],0);   //shift state: unshifted:0
+    All_Vector[1][i][1+1] = KMX_get_KeyValUnderlying_From_KeyCodeUnderlying(keymap,All_Vector[0][i][0],1);   //shift state: shifted:1
   }
-  // ? CHECK if ran OK, vector size is correct -> return 0/1
+
+  return 0;
 }
 
-v_str_2D create_empty_2D( int dim_rows,int dim_shifts)
-{
-  std::string empty = "--";
-  v_str_1D shifts;
-  v_str_2D all;
+bool InitializeGDK(GdkKeymap **keymap,int argc, gchar *argv[]) {
+// get keymap of underlying keyboard
 
-  for ( int i=0; i< dim_rows;i++) {
-    for ( int j=0; j< dim_shifts;j++) {
-      shifts.push_back(empty);
-    }
-    all.push_back(shifts);
-    shifts.clear();
+  gdk_init(&argc, &argv);
+  GdkDisplay *display = gdk_display_get_default();
+  if (!display) {
+    wprintf(L"ERROR: can't get display\n");
+    return 1;
   }
-  //printf("+++++++ dimensions of Vector after create_empty_2D\t\t %li..%li..%li\n", all.size(), all[0].size(),all[1].size());
-  return all;
+
+  *keymap = gdk_keymap_get_for_display(display);
+  if (!keymap) {
+    wprintf(L"ERROR: Can't get keymap\n");
+    gdk_display_close(display);
+    return 2;
+  }
+  return 0;
+}
+
+bool IsKeymanUsedChar(int KV) {
+  //         32            A-Z                      a-z
+  if  ((KV == 0x20 ) || (KV >= 65 && KV <= 90) || (KV >= 97 && KV <= 122) )
+    return true;
+  else
+    return false;
+}
+
+std::u16string convert_DeadkeyValues_To_U16str(int in) {
+
+  if (in == 0 )
+    return u"\0";
+
+  std::string long_name((const char*) gdk_keyval_name (in));                      // e.g. "dead_circumflex" , "U+017F" , "t"
+
+  if ( long_name.substr (0,2) == "U+" )                                           // U+... Unicode value
+    return  CodePointToU16String(in-0x1000000);
+
+  if (in < (int) deadkey_min) {                                                   // no deadkey; no Unicode
```

Comment:
For performance you might want to do this check before you call `gdk_keyval_name`.